### PR TITLE
C++: Add Parameter nodes for functions without bodies

### PR DESCRIPTION
**Pull request by MathiasVP**
This PR fixes an annoyance with `ParameterNode`s in C/C++ dataflow: Parameters of functions without a definition (i.e., when we only have a set of declarations) don't generate `ParameterNode`s.

This is because we define `ParameterNode`s using the `InitializeParameter` instruction in the IR, and without a function body there's no instructions associated with the function.

This PR fixes this by generating new dataflow nodes (from `Parameter`s) and marking those as `ParameterNode`s.

Ideally, only https://github.com/github/codeql/pull/16246/commits/98a3f2da7dd43ca0b82d638fa6b97f2e0b6927bc and https://github.com/github/codeql/pull/16246/commits/06f52c28547f12c7297e52a83f876dfea13e8eff would've been necessary. However, the current setup for marking parameter nodes wasn't very easy to extend to something that wasn't defined in terms of `Instruction`s so it took some refactoring (see https://github.com/github/codeql/pull/16246/commits/111ad8b52583c566fa8b0e9944f6e804b9b81fc2) to allow for this. That commit should be totally behavior preserving.

Commit-by-commit review recommended.

cc @bdrodes

### CODE REVIEW

**Comment by jketema** [comment]
I'm missing something - I think. How do we know that the bodiless parameter nodes are actually only for functions without bodies? I don't see an explicit condition for this anywhere.

**Comment by MathiasVP** [comment]
> I'm missing something - I think. How do we know that the bodiless parameter nodes are actually only for functions without bodies? I don't see an explicit condition for this anywhere.

That's the `not any(InitializeParameterInstruction init).getParameter() = p` part that's in the IPA branch for the new dataflow nodes. At least, that's what I intended when I wrote it 😂

**Comment by jketema** [comment]
> > I'm missing something - I think. How do we know that the bodiless parameter nodes are actually only for functions without bodies? I don't see an explicit condition for this anywhere.
> 
> That's the `not any(InitializeParameterInstruction init).getParameter() = p` part that's in the IPA branch for the new dataflow nodes. At least, that's what I intended when I wrote it 😂

Ah, my brain filtered that out completely because of the horrendous formatting of the diff on my small screen 🤦

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll [hunk 1]
```diff
@@ -767,42 +810,6 @@ class FlowSummaryNode extends Node, TFlowSummaryNode {
...
-  override Declaration getEnclosingCallable() { result = this.getFunction() }
-
-  override Declaration getFunction() { result = init.getEnclosingFunction() }
-
-  /** Gets the underlying operand and the underlying indirection index. */
-  predicate hasInstructionAndIndirectionIndex(Instruction instr, int index) {
-    IndirectInstruction.super.hasInstructionAndIndirectionIndex(instr, index)
-  }
-
-  override Location getLocationImpl() { result = this.getParameter().getLocation() }
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:794]: This no longer occurs after the changes. Is that on purpose?

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:794]: Oh, well spotted. Yeah, there wasn't really a reason to override this as the `getLocationImpl` predicate has a perfectly reasonable implementation from `IndirectInstruction`. In fact, when I replaced this call with `IndirectInstruction.super.getLocationImpl()` ql-for-ql told me that it was redundant so I simply deleted it.

As a sanity check I ran:
```ql
private predicate foo(ParameterNode n) {
  not exists(n.getLocation())
}
```
on a DB I had locally, and got no results back 🎉

**Pull request merged**